### ISSUE TITLE
reassignment to the "throwaway" `_` variable raises false positive `reportAssignmentType` error

### ISSUE THREAD

**Issue by efine-vivodyne** [microsoft/pyright#7665]
**Describe the bug**
Pyright recommends explicitly capturing any return values, and using an underscore as the generic "I don't care about this" variable name if you don't care about it.  Similar for function arguments when you might be overriding something (or otherwise have a required function signature you can't change) and don't care to use a required argument.

However, if these are different types, then pyright throws a error, which seems like a false positive.
```
"dict" is incompatible with "int" (reportAssignmentType)
```

It seems like pyright should recognize that `_` is a special variable name and exclude it from these reports.

**Code or Screenshots**

```python
def lambda_handler(event: dict[str, str], _: int):
    _ = dict() # pretend this is a function that returns some value that you don't care about
```

**VS Code extension or command-line**
command line 1.1.358

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

Pyright does exempt a symbol named `_` from "unreferenced symbol" diagnostics. This makes sense because there is a convention in Python to use `_` to indicate "I don't plan to use this symbol". 

I don't think it's appropriate to exempt _any_ symbols (regardless of their name) from type checking. There's nothing in the Python static typing spec that would indicate this would be appropriate. In your case, you have explicitly declared the type of `_` to be `int`, and a type checker should enforce that declaration.

Pyright never emits diagnostics for unreferenced function parameters, so it's not clear to me why you've chosen to name one of your parameters `_`. It's even a keyword parameter, so its name is significant. Is `_` the appropriate name for this parameter? My recommendation is to choose a more appropriate name for this parameter. Then you can use `_` as an unannotated local variable.

**Comment by efine-vivodyne** [comment]
I don't have control over the function signature (an AWS Lambda function always gets passed two arguments when it is invoked through AWS), but I don't care about the 2nd argument.

If I gave the argument a name that's not `_`, then other linting tools (e.g. ruff) complain that it's an unused argument.

I tried `def lambda_handler(event: dict[str, str], _):`
but that gave
`Type of "lambda_handler" is "(event: dict[str, str], _: Unknown) -> None" (reportUnknownVariableType)`

does that contradict your statement of 
> Pyright never emits diagnostics for unreferenced function parameters

or am I misunderstanding?


I also tried `def lambda_handler(event: dict[str, str], *_):`
but that gave
`Type of "lambda_handler" is partially unknown
    Type of "lambda_handler" is "(event: dict[str, str], *_: Unknown) -> None" (reportUnknownVariableType)`
    


I also tried `def lambda_handler(event: dict[str, str], *):`
but that gave
`error: Keyword parameter must follow "*"`

**Comment by erictraut** [comment]
> ...then other linting tools (e.g. ruff) complain that it's an unused argument...

I don't think linters should complain about an unused function parameter. There are many reasons in python for function parameters to be unused, and this shouldn't be considered an error. Your use case here is one such example, but there are many other legitimate cases. Encouraging developers to rename their function parameters to `_` isn't a good idea. It makes the code less readable, it doesn't work for more than one parameter, and it doesn't recognize the fact that parameter names are significant in Python.

I presume that this rule is configurable in ruff. If so, I recommend configuring it to not generate a linting error in this case. You might also want to report an issue to the maintainers of ruff to see if they will exempt function parameters from this check by default.

> does that contradict your statement?

No, I'm referring to pyright's `reportUnusedVariable` diagnostic check. It always exempts function parameters for the reasons I mention above.

> I also tried... `*`

That won't work here. The `*` has a specific meaning in a python function signature. It was introduced in [PEP 570](https://peps.python.org/pep-0570/).
